python/cudf/cudf/core/dataframe.py

```diff
@@ -1336,32 +1340,35 @@ def reset_index(self, drop=False, inplace=False):
         if inplace is True:
             for column_name in set(out.columns) - set(self.columns):
                 self[column_name] = out[column_name]
-                self._cols.move_to_end(column_name, last=False)
-            self._index = RangeIndex(len(self))
+                self._data.move_to_end(column_name, last=False)
+            self.index = RangeIndex(len(self))
         else:
             return out.set_index(RangeIndex(len(self)))
 
     def take(self, positions, ignore_index=False):
-        out = DataFrame()
-        if self._cols:
-            for i, col_name in enumerate(self._cols.keys()):
-                out[col_name] = self[col_name][positions]
         if ignore_index:
-            out.index = RangeIndex(len(out))
+            index = RangeIndex(len(positions))
         else:
-            out._index = self.index.take(positions)
-        return out
+            index = self.index.take(positions)
+        out = DataFrame()
+        if self._data:
+            for i, col_name in enumerate(self._data.keys()):
+                out[col_name] = self._data[col_name][positions]
+        return out.set_index(index)
 
     def _take_columns(self, positions):
         positions = Series(positions)
-        column_names = list(self._cols.keys())
-        column_values = list(self._cols.values())
+        column_names = list(self._data.keys())
+        column_values = list(self._data.values())
         result = DataFrame()
         for idx in range(len(positions)):
-            result[column_names[positions[idx]]] = column_values[
-                positions[idx]
-            ]
-        result._index = self._index
+            if len(self) == 0:
+                result[idx] = cudf.Series([])
```

Comment:
use empty column constructor instead of Series constructor here?

Reply:
Done